reframe/frontend/statistics.py

```diff
@@ -1,89 +1,98 @@
+from itertools import chain
+
 import reframe.core.debug as debug
+import reframe.core.exceptions as exc
 
 
 class TestStats:
     """Stores test case statistics."""
 
-    def __init__(self, tasks=[]):
-        if not hasattr(tasks, '__iter__'):
-            raise TypeError('expected an iterable')
-
-        # Store test cases per partition internally
-        self._tasks_bypart = {}
-        for t in tasks:
-            partition = t.check.current_partition
-            partname = partition.fullname if partition else 'None'
-            tclist = self._tasks_bypart.setdefault(partname, [])
-            tclist.append(t)
+    def __init__(self):
+        self._tasks = []
+        self._try_num = -1
 
     def __repr__(self):
         return debug.repr(self)
 
-    def num_failures(self, partition=None):
-        num_fails = 0
-        if partition:
-            num_fails += len([
-                t for t in self._tasks_bypart[partition] if t.failed
-            ])
-        else:
-            # count all failures
-            for tclist in self._tasks_bypart.values():
-                num_fails += len([t for t in tclist if t.failed])
-
-        return num_fails
-
-    def num_failures_stage(self, stage):
-        num_fails = 0
-        for tclist in self._tasks_bypart.values():
-            num_fails += len([t for t in tclist if t.failed_stage == stage])
-
-        return num_fails
-
-    def num_cases(self, partition=None):
-        num_cases = 0
-        if partition:
-            num_cases += len(self._tasks_bypart[partition])
-        else:
-            # count all failures
-            for tclist in self._tasks_bypart.values():
-                num_cases += len(tclist)
-
-        return num_cases
-
-    def failure_report(self):
+    def next_try(self, try_num):
+        # TODO: RAISE DIFFERENT error?
+        if try_num != self._try_num + 1:
+            raise exc.ReframeFatalError('try_num out of sync')
+
+        self._try_num = try_num
+        self._tasks.append([])
+
+    def add_task(self, task):
+        self._tasks[self._try_num].append(task)
+
+    # The try_num can be indexed from the end (e.g. -1 for the last retry)
+    def _convert_try_num(self, try_num):
```

Comment:
I don't see the purpose of this method. Python supports backward indexing with negative indexes natively.

Reply:
Yes, with the new design, it is not anymore needed everywhere, but only needed in `failure_report()`. Adapted.